### PR TITLE
[Draft] Add per-tensor KV cache quantization

### PR DESCRIPTION
**Pull request by jrplatin**
# Description

TODO

# Accuracy / Performance

Please describe how you tested this change, and include any instructions and/or
commands to reproduce.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [X] I have performed a self-review of my code.
- [X] I have necessary comments in my code, particularly in hard-to-understand areas.
- [X] I have made or will make corresponding changes to any relevant documentation.

### CODE REVIEW

File: tpu_commons/kernels/ragged_paged_attention/kernel.py [hunk 1]
```diff
@@ -237,8 +239,10 @@ def static_validate_inputs(
     _, num_q_heads, head_dim = q.shape
     _, _, num_combined_kv_heads, head_dim_k = kv_pages.shape
     assert num_combined_kv_heads % 2 == 0
-    assert isinstance(k_scale, float) or k_scale is None
-    assert isinstance(v_scale, float) or v_scale is None
+    assert (isinstance(k_scale, jax.Array)
```
Comment (yaochengji) [inline tpu_commons/kernels/ragged_paged_attention/kernel.py:242]: validate k_scale and v_scale's dtype?

File: tpu_commons/kernels/ragged_paged_attention/kernel.py [hunk 2]
```diff
@@ -740,8 +758,8 @@ def ragged_paged_attention(
     sliding_window: the sliding window size for the attention.
     soft_cap: the logit soft cap for the attention.
     mask_value: mask value for causal mask.
-    k_scale: the scale for the key cache.
-    v_scale: the scale for the value cache.
+    k_scale: the scale for the key cache.  Should be a jax.Array of shape (1,).
```
Comment (bythew3i) [inline tpu_commons/kernels/ragged_paged_attention/kernel.py:761]: We should make a kv_scale array (2, ) to save some smem.... because each (1,) is actually taking (128,) space.

File: tpu_commons/kernels/ragged_paged_attention/kernel.py [hunk 3]
```diff
@@ -289,6 +293,8 @@ def ragged_paged_attention_kernel(
     seq_buf_idx_ref,
     # TODO(jevinjiang): if OOM in SMEM, consider pack to other scalar refs.
     num_seqs_ref,
+    k_scale_ref,
+    v_scale_ref,
```
Comment (bythew3i) [inline tpu_commons/kernels/ragged_paged_attention/kernel.py:297]: We should pass kv_scale_ref as scalar prefetch.

File: tpu_commons/kernels/ragged_paged_attention/kernel.py [hunk 4]
```diff
@@ -120,10 +122,10 @@ def ref_ragged_paged_attention(
         v = kv_pages[indices, :, 1::2, :].reshape(-1, num_kv_heads,
                                                   head_dim)[:kv_len]
         if k_scale is not None:
-            k = k.astype(jnp.float32) * k_scale
+            k = k.astype(jnp.float32) * jnp.broadcast_to(k_scale, k.shape)
```
Comment (bythew3i) [inline tpu_commons/kernels/ragged_paged_attention/kernel.py:125]: this is expensive because it uses XLU... we should load scale from SMEM ref

Comment (jrplatin) [reply tpu_commons/kernels/ragged_paged_attention/kernel.py:125]: This is in the reference ragged paged attention -- does it matter?

File: tpu_commons/kernels/ragged_paged_attention/kernel.py (hunk 3 above)
Comment (yaochengji) [reply tpu_commons/kernels/ragged_paged_attention/kernel.py:297]: I remember scalar prefetch cannot be multiplied with the k and v from vmem. But @jrplatin feel free to have another try.

Comment (bythew3i) [reply tpu_commons/kernels/ragged_paged_attention/kernel.py:297]: It will work after you load to SREG.

File: tpu_commons/kernels/ragged_paged_attention/kernel.py [hunk 5]
```diff
@@ -489,9 +495,15 @@ def load_with_init(ref, init_val):
...
-                "nd,md->nm", q, k, preferred_element_type=jnp.float32) *
-                  sm_scale)
+            einsum = functools.partial(jnp.einsum,
+                                       preferred_element_type=jnp.float32)
+            if k_scale is not None:
+                k = qarray.QArray(k,
+                                  k_scale.reshape(1, 1),
+                                  zero_point=None,
+                                  qtype=k.dtype)
+                einsum = qwix_einsum
```
Comment (yaochengji) [inline tpu_commons/kernels/ragged_paged_attention/kernel.py:505]: Why do we use qwix_einsum here?
